### PR TITLE
fs: Allocate file/socket dynamically

### PR DESCRIPTION
**Pull request by Donny9**
## Summary
Allocate file/socket dynamically.
## Impact
Dynamically allocated file descriptor for every task can reduce ram, especially super task in a project.
## Testing
daily test

### CODE REVIEW

File: sched/Kconfig [hunk 1]
```diff
@@ -1058,12 +1058,12 @@ config SDCLONE_DISABLE
 	descriptors by task_create() when a new task is started. If
 	set, all sockets will appear to be closed in the new task.
 
-config NFILE_DESCRIPTORS
-	int "Maximum number of file descriptors per task"
-	default 16
+config NFCHUNK_DESCRIPTORS
```
Comment (acassis) [inline sched/Kconfig:1061]: I think the name NFCHUNK_DESCRIPTORS is enigmatic, why not NFILE_DESCRIPTORS_ROW or something similar?

**Comment by davids5** [comment]
@Donny9  is there a set of PRs for this that need to come in in some order?

File: sched/Kconfig (hunk 1 above)
Comment (Donny9) [reply sched/Kconfig:1061]: the  mean isn't NFILE_DESCRIPTORS_ROW, you can see more in #2834. we need  config the size of each chunk in the structure of  dobule fd array. But, this only my suggest.

**Comment by Donny9** [comment]
> @Donny9 is there a set of PRs for this that need to come in in some order?

yes, This patch associates with apache/incubator-nuttx-apps#609，it need to merge firstly.

File: sched/Kconfig (hunk 1 above)
Comment (xiaoxiang781216) [reply sched/Kconfig:1061]: This option control how many file handles to extend in each reallocation, NFILE_DESCRIPTORS_ROW obviously isn't a good name, @acassis do you have other better suggestion?

**Comment by xiaoxiang781216** [comment]
depends on https://github.com/apache/incubator-nuttx-apps/pull/612

File: sched/Kconfig (hunk 1 above)
Comment (davids5) [reply sched/Kconfig:1061]: Will `NFILE_DESCRIPTORS_PER_ALLOC` work?

Comment (Donny9) [reply sched/Kconfig:1061]: NFILE_DESCRIPTORS_PER_BLOCK?
NFILE_DESCRIPTORS_PER_CHUNK?

Comment (xiaoxiang781216) [reply sched/Kconfig:1061]: PER_BLOCK/PER_CHUNK looks better, @davids5 @acassis what's your opinion?

**Review by xiaoxiang781216 — APPROVED** [review]

File: sched/Kconfig (hunk 1 above)
Comment (davids5) [reply sched/Kconfig]: +1 for NFILE_DESCRIPTORS_PER_BLOCK

**Comment by xiaoxiang781216** [comment]
It seems that CI cancel "Build Documentation" accidently, all other check finish without any problem. So let's ignore the false alarm.

**Comment by masayuki2009** [comment]
@xiaoxiang781216 @Donny9 

Hmm, it seems that telnetd can not accept an incoming connection again.
Could you please check lm3s6965-ek:discover with QEMU?

**Comment by Donny9** [comment]
> @xiaoxiang781216 @Donny9
> 
> Hmm, it seems that telnetd can not accept an incoming connection again.
> Could you please check lm3s6965-ek:discover with QEMU?

Okay. I check it

**Comment by Donny9** [comment]
> @xiaoxiang781216 @Donny9
> 
> Hmm, it seems that telnetd can not accept an incoming connection again.
> Could you please check lm3s6965-ek:discover with QEMU?

@masayuki2009  please apply this patch to fix this issue: https://github.com/apache/incubator-nuttx-apps/pull/619. Thank you very much.

**Pull request merged**